File: src/converters/FormatConverter.ts
/**
 * Format Converter - Quantum chemistry file format conversion adapter
 *
 * This module provides a TypeScript interface to the Python dpdata-based
 * format conversion backend. It supports conversion between VASP, Gaussian,
 * ORCA, XYZ, PDB, CIF, and other quantum chemistry file formats.
 */

import * as vscode from 'vscode';
import * as path from 'path';
import { spawn } from 'child_process';

/**
 * Supported file formats for conversion
 */
export enum SupportedFormat {
  VASP = 'vasp',
  POSCAR = 'poscar',
  CONTCAR = 'contcar',
  Gaussian = 'gaussian',
  GJF = 'gjf',
  COM = 'com',
  ORCA = 'orca',
  INP = 'inp',
  XYZ = 'xyz',
  PDB = 'pdb',
  CIF = 'cif',
  CP2K = 'cp2k',
  QE = 'qe',
  QuantumEspresso = 'quantum_espresso',
}

/**
 * Conversion metadata
 */
export interface ConversionMetadata {
  natoms?: number;
  nbonds?: number;
  elements?: string[];
  atom_counts?: number[];
  frames_count?: number;
  original_format?: string;
}

/**
 * Result of a format conversion operation
 */
export interface ConversionResult {
  success: boolean;
  input_format?: string;
  output_format?: string;
  atoms_count?: number;
  frames_count?: number;
  metadata?: ConversionMetadata;
  output_file?: string;
  error?: string;
  error_type?: string;
}

/**
 * Batch conversion result
 */
export interface BatchConversionResult {
  success: boolean;
  total: number;
  successful: number;
  failed: number;
  results: Array<{
    input_file: string;
    output_file: string;
    result: ConversionResult;
  }>;
}

/**
 * Format detection result
 */
export interface FormatDetection {
  format: SupportedFormat;
  confidence: number;
  extension: string;
}

/**
 * Configuration for the FormatConverter
 */
export interface FormatConverterConfig {
  pythonPath?: string;
  scriptPath?: string;
  preserveMetadata?: boolean;
}

/**
 * Main format converter class
 */
export class FormatConverter {
  private config: FormatConverterConfig;
  private outputChannel: vscode.OutputChannel;

  constructor(config: FormatConverterConfig = {}) {
    this.config = {
      pythonPath: config.pythonPath || 'python3',
      scriptPath:
        config.scriptPath || path.resolve(__dirname, '../../../python/format_converter.py'),
      preserveMetadata: config.preserveMetadata ?? true,
    };
    this.outputChannel = vscode.window.createOutputChannel('OpenQC Format Converter');
  }

  /**
   * Detect the format of a file from its extension and content
   */
  static detectFormat(filePath: string): FormatDetection {
    const ext = path.extname(filePath).toLowerCase();
    const basename = path.basename(filePath).toLowerCase();

    // Special VASP files
    if (basename === 'poscar') {
      return { format: SupportedFormat.POSCAR, confidence: 1.0, extension: '' };
    }
    if (basename === 'contcar') {
      return { format: SupportedFormat.CONTCAR, confidence: 1.0, extension: '' };
    }
    if (basename === 'incar') {
      return { format: SupportedFormat.VASP, confidence: 1.0, extension: '' };
    }
    if (basename === 'kpoints') {
      return { format: SupportedFormat.VASP, confidence: 1.0, extension: '' };
    }
    if (basename === 'potcar') {
      return { format: SupportedFormat.VASP, confidence: 1.0, extension: '' };
    }

    // Extension-based detection
    const formatMap: Record<string, SupportedFormat> = {
      '.xyz': SupportedFormat.XYZ,
      '.pdb': SupportedFormat.PDB,
      '.cif': SupportedFormat.CIF,
      '.gjf': SupportedFormat.GJF,
      '.com': SupportedFormat.COM,
      '.inp': SupportedFormat.INP,
    };

    const format = formatMap[ext];
    if (format) {
      return { format, confidence: 0.9, extension: ext };
    }

    // Default to VASP
    return { format: SupportedFormat.POSCAR, confidence: 0.5, extension: ext };
  }

  /**
   * Convert a single file from one format to another
   *
   * @param inputPath Path to input file
   * @param outputPath Path to output file
   * @param fromFormat Source format (auto-detected if not specified)
   * @param toFormat Target format (inferred from output path if not specified)
   */
  async convert(
    inputPath: string,
    outputPath: string,
    fromFormat?: SupportedFormat,
    toFormat?: SupportedFormat
  ): Promise<ConversionResult> {
    try {
      this.log(`Converting ${inputPath} to ${outputPath}`);

      const detection = fromFormat ? undefined : FormatConverter.detectFormat(inputPath);
      const sourceFormat = fromFormat || detection?.format;
      const targetFormat = toFormat || FormatConverter.detectFormat(outputPath).format;

      const args = [this.config.scriptPath!, inputPath, outputPath];

      if (fromFormat) {
        args.push('--from', fromFormat);
      }
      if (toFormat) {
        args.push('--to', toFormat);
      }
      if (!this.config.preserveMetadata) {
        args.push('--no-metadata');
      }
      args.push('--json');

      const stdout = await this.runPython(args);
      const result = JSON.parse(stdout) as ConversionResult;

      if (result.success) {
        this.log(`Conversion successful: ${result.output_file}`);
      } else {
        this.logError(`Conversion failed: ${result.error}`);
      }

      return result;
    } catch (error) {
      const errorMessage = error instanceof Error ? error.message : String(error);
      this.logError(`Conversion error: ${errorMessage}`);
      return {
        success: false,
        error: errorMessage,
        error_type: 'ExecutionError',
      };
    }
  }

  /**
   * Batch convert multiple files
   *
   * @param inputPaths Array of input file paths
   * @param outputDir Output directory
   * @param toFormat Target format
   * @param fromFormat Source format (auto-detected if not specified)
   */
  async batchConvert(
    inputPaths: string[],
    outputDir: string,
    toFormat: SupportedFormat,
    fromFormat?: SupportedFormat
  ): Promise<BatchConversionResult> {
    try {
      this.log(`Batch converting ${inputPaths.length} files to ${toFormat}`);

      const args = [
        this.config.scriptPath!,
        '--batch',
        ...inputPaths,
        '--to',
        toFormat,
        '--output-dir',
        outputDir,
        '--json',
      ];

      if (fromFormat) {
        args.push('--from', fromFormat);
      }
      if (!this.config.preserveMetadata) {
        args.push('--no-metadata');
      }

      const stdout = await this.runPython(args);
      const result = JSON.parse(stdout) as BatchConversionResult;

      this.log(`Batch conversion complete: ${result.successful}/${result.total} successful`);

      return result;
    } catch (error) {
      const errorMessage = error instanceof Error ? error.message : String(error);
      this.logError(`Batch conversion error: ${errorMessage}`);
      return {
        success: false,
        total: inputPaths.length,
        successful: 0,
        failed: inputPaths.length,
        results: [],
      };
    }
  }

  /**
   * Convert the current editor document to a different format
   */
  async convertCurrentDocument(
    targetFormat: SupportedFormat
  ): Promise<ConversionResult | undefined> {
    const editor = vscode.window.activeTextEditor;
    if (!editor) {
      vscode.window.showErrorMessage('No active document');
      return undefined;
    }

    const inputPath = editor.document.uri.fsPath;
    const inputDir = path.dirname(inputPath);
    const inputBasename = path.basename(inputPath, path.extname(inputPath));
    const extension = this.getExtensionForFormat(targetFormat);
    const outputPath = path.join(inputDir, `${inputBasename}.${extension}`);

    return this.convert(inputPath, outputPath, undefined, targetFormat);
  }

  /**
   * Get file extension for a given format
   */
  getExtensionForFormat(format: SupportedFormat): string {
    const extMap: Record<SupportedFormat, string> = {
      [SupportedFormat.VASP]: 'POSCAR',
      [SupportedFormat.POSCAR]: 'POSCAR',
      [SupportedFormat.CONTCAR]: 'CONTCAR',
      [SupportedFormat.Gaussian]: 'gjf',
      [SupportedFormat.GJF]: 'gjf',
      [SupportedFormat.COM]: 'com',
      [SupportedFormat.ORCA]: 'inp',
      [SupportedFormat.INP]: 'inp',
      [SupportedFormat.XYZ]: 'xyz',
      [SupportedFormat.PDB]: 'pdb',
      [SupportedFormat.CIF]: 'cif',
      [SupportedFormat.CP2K]: 'inp',
      [SupportedFormat.QE]: 'in',
      [SupportedFormat.QuantumEspresso]: 'in',
    };
    return extMap[format];
  }

  /**
   * Check if the Python backend is available
   */
  async checkBackend(): Promise<boolean> {
    try {
      await this.runPython(['--version']);
      await this.runPython(['-c', 'import dpdata']);
      return true;
    } catch {
      return false;
    }
  }

  /**
   * Show setup instructions if backend is not available
   */
  async showSetupInstructions(): Promise<void> {
    const action = await vscode.window.showWarningMessage(
      'OpenQC Format Converter requires Python and dpdata. Install now?',
      'Install',
      'Cancel'
    );

    if (action === 'Install') {
      const terminal = vscode.window.createTerminal('OpenQC Setup');
      terminal.sendText('pip install dpdata');
      terminal.show();
    }
  }

  /**
   * Quick conversion using the existing StructureConverter
   * (for XYZ output without Python backend)
   */
  static convertToXYZ(
    atoms: Array<{ elem: string; x: number; y: number; z: number }>,
    comment: string = 'molecule'
  ): string {
    const lines: string[] = [];

    lines.push(String(atoms.length));
    lines.push(comment);

    for (const atom of atoms) {
      const x = atom.x.toFixed(6);
      const y = atom.y.toFixed(6);
      const z = atom.z.toFixed(6);
      lines.push(`${atom.elem} ${x} ${y} ${z}`);
    }

    return lines.join('\n');
  }

  private log(message: string): void {
    this.outputChannel.appendLine(message);
  }

  private logError(message: string): void {
    this.outputChannel.appendLine(`[ERROR] ${message}`);
  }

  private runPython(args: string[]): Promise<string> {
    return new Promise((resolve, reject) => {
      const child = spawn(this.config.pythonPath!, args, { shell: false });
      let stdout = '';
      let stderr = '';

      child.stdout.on('data', chunk => {
        stdout += chunk.toString();
      });

      child.stderr.on('data', chunk => {
        stderr += chunk.toString();
      });

      child.on('error', reject);
      child.on('close', code => {
        if (code === 0) {
          resolve(stdout);
          return;
        }

        reject(new Error(stderr.trim() || `Python process exited with code ${code}`));
      });
    });
  }

  dispose(): void {
    this.outputChannel.dispose();
  }
}

/**
 * Helper function to create and show a quick format conversion
 */
export async function quickConvert(targetFormat: SupportedFormat): Promise<void> {
  const converter = new FormatConverter();

  if (!(await converter.checkBackend())) {
    await converter.showSetupInstructions();
    return;
  }

  const result = await converter.convertCurrentDocument(targetFormat);

  if (result?.success) {
    const action = await vscode.window.showInformationMessage(
      `Successfully converted to ${targetFormat.toUpperCase()}`,
      'Open File',
      'Copy to Clipboard'
    );

    if (action === 'Open File' && result.output_file) {
      const doc = await vscode.workspace.openTextDocument(result.output_file);
      await vscode.window.showTextDocument(doc);
    } else if (action === 'Copy to Clipboard') {
      // Read and copy file content
      const content = await vscode.workspace.fs.readFile(vscode.Uri.file(result.output_file!));
      await vscode.env.clipboard.writeText(Buffer.from(content).toString('utf8'));
    }
  } else {
    vscode.window.showErrorMessage(`Conversion failed: ${result?.error}`);
  }
}

File: python/format_converter.py
#!/usr/bin/env python3
"""
OpenQC Format Converter - Python backend for quantum chemistry file format conversion.

This module uses dpdata library to convert between various quantum chemistry file formats.
Supported formats: VASP, Gaussian, ORCA, XYZ, PDB, CIF, and more.
"""

import sys
import json
import argparse
import logging
from pathlib import Path
from typing import Dict, List, Any, Optional

logger = logging.getLogger(__name__)


def convert_format(
    input_file: str,
    output_file: str,
    from_format: Optional[str] = None,
    to_format: Optional[str] = None,
    preserve_metadata: bool = True,
) -> Dict[str, Any]:
    """
    Convert quantum chemistry file between different formats.

    Args:
        input_file: Path to input file
        output_file: Path to output file
        from_format: Source format (auto-detected if None)
        to_format: Target format (inferred from output extension if None)
        preserve_metadata: Whether to preserve calculation metadata

    Returns:
        Dictionary with conversion result and metadata
    """
    try:
        import dpdata
    except ImportError:
        return {
            "success": False,
            "error": "dpdata library not found. Install with: pip install dpdata",
            "error_type": "ImportError",
        }

    input_path = Path(input_file)
    output_path = Path(output_file)

    if not input_path.exists():
        return {
            "success": False,
            "error": f"Input file not found: {input_file}",
            "error_type": "FileNotFoundError",
        }

    try:
        # Auto-detect input format from file extension if not specified
        if from_format is None:
            from_format = _detect_format(input_path)

        # Determine output format from extension if not specified
        if to_format is None:
            to_format = _detect_format(output_path)

        # Load system using dpdata
        system = _load_system(input_path, from_format)

        if system is None:
            return {
                "success": False,
                "error": f"Failed to load input file with format: {from_format}",
                "error_type": "LoadError",
            }

        # Extract metadata if requested
        metadata = {}
        if preserve_metadata:
            metadata = _extract_metadata(system, from_format)

        # Convert and save to output format
        _save_system(system, output_path, to_format)

        return {
            "success": True,
            "input_format": from_format,
            "output_format": to_format,
            "atoms_count": system.get_natoms(),
            "frames_count": len(system),
            "metadata": metadata,
            "output_file": str(output_path.absolute()),
        }

    except Exception as e:
        return {
            "success": False,
            "error": str(e),
            "error_type": type(e).__name__,
        }


def batch_convert(
    input_files: List[str],
    output_dir: str,
    to_format: str,
    from_format: Optional[str] = None,
    preserve_metadata: bool = True,
) -> Dict[str, Any]:
    """
    Batch convert multiple files to a target format.

    Args:
        input_files: List of input file paths
        output_dir: Directory for output files
        to_format: Target format
        from_format: Source format (auto-detected if None)
        preserve_metadata: Whether to preserve metadata

    Returns:
        Dictionary with batch conversion results
    """
    results = []
    output_path = Path(output_dir)
    output_path.mkdir(parents=True, exist_ok=True)

    for input_file in input_files:
        input_path = Path(input_file)
        output_file = output_path / f"{input_path.stem}.{_get_extension(to_format)}"

        result = convert_format(
            input_file=str(input_path),
            output_file=str(output_file),
            from_format=from_format,
            to_format=to_format,
            preserve_metadata=preserve_metadata,
        )

        results.append({
            "input_file": input_file,
            "output_file": str(output_file),
            "result": result,
        })

    successful = sum(1 for r in results if r["result"].get("success"))
    total = len(results)

    return {
        "success": successful == total,
        "total": total,
        "successful": successful,
        "failed": total - successful,
        "results": results,
    }


def supported_formats() -> Dict[str, List[str]]:
    """
    Get list of supported formats for conversion.

    Returns:
        Dictionary with input and output format lists
    """
    try:
        import dpdata
        # dpdata supports various formats
        return {
            "input": [
                "vasp", "poscar", "contcar",
                "gaussian", "gjf", "com",
                "orca", "inp",
                "xyz",
                "pdb",
                "cif",
                "cp2k",
                "qe", "quantum_espresso", "pw_in",
            ],
            "output": [
                "vasp", "poscar",
                "gaussian", "gjf", "com",
                "orca", "inp",
                "xyz",
                "pdb",
                "cif",
            ],
        }
    except ImportError:
        return {"input": [], "output": []}


def _detect_format(file_path: Path) -> str:
    """Auto-detect format from file extension and name."""
    filename = file_path.name.lower()
    ext = file_path.suffix.lower()

    # Special VASP files
    if filename == "poscar" or filename == "contcar":
        return "vasp/poscar"
    if filename == "incar":
        return "vasp/incar"
    if filename == "kpoints":
        return "vasp/kpoints"
    if filename == "potcar":
        return "vasp/potcar"

    # Extension-based detection
    format_map = {
        ".xyz": "xyz",
        ".pdb": "pdb",
        ".cif": "cif",
        ".gjf": "gaussian",
        ".com": "gaussian",
        ".inp": "orca",
    }

    return format_map.get(ext, "vasp/poscar")


def _get_extension(format_name: str) -> str:
    """Get file extension for a format."""
    ext_map = {
        "vasp": "poscar",
        "poscar": "poscar",
        "gaussian": "gjf",
        "gjf": "gjf",
        "com": "com",
        "orca": "inp",
        "inp": "inp",
        "xyz": "xyz",
        "pdb": "pdb",
        "cif": "cif",
    }
    return ext_map.get(format_name.lower(), "out")


def _load_system(file_path: Path, format_name: str):
    """Load system using dpdata."""
    import dpdata

    format_lower = format_name.lower()

    if "vasp" in format_lower or "poscar" in format_lower:
        return dpdata.System(str(file_path), fmt='vasp/poscar')
    elif "gaussian" in format_lower:
        return dpdata.System(str(file_path), fmt='gaussian')
    elif "orca" in format_lower:
        return dpdata.System(str(file_path), fmt='orca')
    elif "xyz" in format_lower:
        return dpdata.System(str(file_path), fmt='xyz')
    elif "pdb" in format_lower:
        return dpdata.System(str(file_path), fmt='pdb')
    elif "cif" in format_lower:
        return dpdata.System(str(file_path), fmt='cif')
    elif "cp2k" in format_lower:
        return dpdata.System(str(file_path), fmt='cp2k')
    else:
        # Try auto-detection
        return dpdata.System(str(file_path))


def _save_system(system, file_path: Path, format_name: str) -> None:
    """Save system using dpdata."""
    format_lower = format_name.lower()

    if "vasp" in format_lower or "poscar" in format_lower:
        system.to_vasp_poscar(str(file_path))
    elif "gaussian" in format_lower:
        system.to_gaussian_log(str(file_path))
    elif "xyz" in format_lower:
        system.to_xyz(str(file_path))
    elif "pdb" in format_lower:
        system.to_pdb(str(file_path))
    elif "cif" in format_lower:
        system.to_cif(str(file_path))
    else:
        # Default to POSCAR
        system.to_vasp_poscar(str(file_path))


def _extract_metadata(system, format_name: str) -> Dict[str, Any]:
    """Extract metadata from the system."""
    metadata = {
        "natoms": int(system.get_natoms()),
        "nbonds": int(getattr(system, "nbonds", 0)) if hasattr(system, "nbonds") else None,
    }

    # Add element information
    if hasattr(system, "data"):
        data = system.data
        if "atom_names" in data:
            metadata["elements"] = list(data["atom_names"])
        if "atom_numbs" in data:
            # Convert numpy int64 to regular int for JSON serialization
            metadata["atom_counts"] = [int(x) for x in data["atom_numbs"]]

    # Format-specific metadata
    if "vasp" in format_name.lower():
        metadata["original_format"] = "VASP"
    elif "gaussian" in format_name.lower():
        metadata["original_format"] = "Gaussian"
    elif "orca" in format_name.lower():
        metadata["original_format"] = "ORCA"

    return metadata


def main():
    """CLI entry point."""
    parser = argparse.ArgumentParser(
        description="Convert quantum chemistry file formats",
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog="""
Examples:
  # Convert VASP POSCAR to XYZ
  python format_converter.py POSCAR output.xyz

  # Convert Gaussian to ORCA
  python format_converter.py input.gjf output.inp --from gaussian --to orca

  # Batch convert multiple files
  python format_converter.py --batch file1.xyz file2.xyz --to vasp --output-dir ./converted

  # List supported formats
  python format_converter.py --list-formats
        """
    )

    parser.add_argument("input", nargs="?", help="Input file path")
    parser.add_argument("output", nargs="?", help="Output file path")
    parser.add_argument("--from", dest="from_format", help="Input format")
    parser.add_argument("--to", dest="to_format", help="Output format")
    parser.add_argument("--batch", nargs="+", help="Batch convert multiple files")
    parser.add_argument("--output-dir", help="Output directory for batch conversion")
    parser.add_argument("--no-metadata", action="store_true", help="Don't preserve metadata")
    parser.add_argument("--list-formats", action="store_true", help="List supported formats")
    parser.add_argument("--json", action="store_true", help="Output result as JSON")

    args = parser.parse_args()

    if args.list_formats:
        formats = supported_formats()
        print("Supported input formats:", ", ".join(formats["input"]))
        print("Supported output formats:", ", ".join(formats["output"]))
        return 0

    if args.batch:
        if not args.to_format:
            logger.error("--to format required for batch conversion")
            print("Error: --to format required for batch conversion", file=sys.stderr)
            return 1

        result = batch_convert(
            input_files=args.batch,
            output_dir=args.output_dir or "./converted",
            to_format=args.to_format,
            from_format=args.from_format,
            preserve_metadata=not args.no_metadata,
        )
    else:
        if not args.input or not args.output:
            parser.print_help()
            return 1

        result = convert_format(
            input_file=args.input,
            output_file=args.output,
            from_format=args.from_format,
            to_format=args.to_format,
            preserve_metadata=not args.no_metadata,
        )

    if args.json or args.batch:
        print(json.dumps(result, indent=2))
    else:
        if result.get("success"):
            logger.info(f"Successfully converted to {result['output_format']} format")
            print(f"Successfully converted to {result['output_format']} format")
            print(f"Output: {result.get('output_file')}")
            print(f"Atoms: {result.get('atoms_count')}")
        else:
            logger.error(f"Conversion failed: {result.get('error')}")
            print(f"Error: {result.get('error')}", file=sys.stderr)
            return 1

    return 0


if __name__ == "__main__":
    sys.exit(main())
